src/Bicep.Core/Semantics/Namespaces/ThisNamespaceType.cs

```diff
@@ -0,0 +1,107 @@
+
+    public static NamespaceSettings Settings { get; } = new(
+        IsSingleton: true,
+        BicepExtensionName: BuiltInName,
+        ConfigurationType: null,
+        TemplateExtensionName: "This",
+        TemplateExtensionVersion: "1.0.0");
+
+    private static FunctionResult GetExistsReturnResult(SemanticModel model, IDiagnosticWriter diagnostics, FunctionCallSyntaxBase functionCall, ImmutableArray<TypeSymbol> argumentTypes)
+    {
+        // Compile to not(empty(target('full')))
+        var targetExpression = new FunctionCallExpression(functionCall, "target", [new StringLiteralExpression(functionCall, "full")]);
+        var emptyExpression = new FunctionCallExpression(functionCall, "empty", [targetExpression]);
+        var notExpression = new FunctionCallExpression(functionCall, "not", [emptyExpression]);
+
+        return new(LanguageConstants.Bool, notExpression);
+    }
+
+    private static FunctionResult GetExistingResourceReturnResult(SemanticModel model, IDiagnosticWriter diagnostics, FunctionCallSyntaxBase functionCall, ImmutableArray<TypeSymbol> argumentTypes)
+    {
+        var derived = TryGetExistingResourceType(model, functionCall) ?? LanguageConstants.Object;
+        var resourceType = TryGetEnclosingResourceType(model, functionCall);
+
+        // For extensible resources, use target() without 'full' since they don't have the standard Azure resource structure
+        var isExtensibleResource = resourceType is not null && !resourceType.IsAzResource();
+        var targetExpression = isExtensibleResource
+            ? new FunctionCallExpression(functionCall, "target", [])
+            : new FunctionCallExpression(functionCall, "target", [new StringLiteralExpression(functionCall, "full")]);
+
+        return new(derived, targetExpression);
+    }
+
+    private static ResourceType? TryGetEnclosingResourceType(SemanticModel model, SyntaxBase syntax)
+    {
+        var decl = model.Binder.GetNearestAncestor<ResourceDeclarationSyntax>(syntax);
+        if (decl is null)
+        {
+            return null;
+        }
+
+        if (model.GetDeclaredType(decl) is ResourceType rt)
+        {
+            return rt;
+        }
+
+        return null;
+    }
+
+    private static TypeSymbol? TryGetExistingResourceType(SemanticModel model, SyntaxBase syntax)
+    {
+        var resourceType = TryGetEnclosingResourceType(model, syntax);
+        if (resourceType is null)
+        {
+            return null;
+        }
+
+        return resourceType.Body.Type;
+    }
+
+    private static IEnumerable<FunctionOverload> GetThisFunctions()
+    {
+        yield return new FunctionOverloadBuilder("exists")
+            .WithReturnType(LanguageConstants.Bool)
+            .WithReturnResultBuilder(GetExistsReturnResult, LanguageConstants.Bool)
+            .WithGenericDescription("Returns whether the current resource exists.")
+            .WithFlags(FunctionFlags.RequiresInlining)
```

Comment:
[Nit]: I think this flag won't do anything since `this` is only defined in syntactic contexts that disallow statements (i.e., resource body expressions).